fvm/evm/precompiles/abi.go

```diff
@@ -0,0 +1,223 @@
+package precompiles
+
+import (
+	"encoding/binary"
+	"errors"
+	"math/big"
+
+	gethCommon "github.com/ethereum/go-ethereum/common"
+)
+
+// This package provides fast and effient
+// utilities needed for abi encoding and decoding
+// encodings are mostly used for testing purpose
+// if more complex encoding and decoding is needed please
+// use the abi package and pass teh ABIs, though
+// that has a performance overhead.
+const (
+	FixedSizeUnitDataReadSize = 32
+	Bytes4DataReadSize        = 4
+	Bytes8DataReadSize        = 8
+	Bytes32DataReadSize       = 32
+	Uint64ByteSize            = 8
+
+	EncodedBoolSize    = FixedSizeUnitDataReadSize
+	EncodedAddressSize = FixedSizeUnitDataReadSize
+	EncodedBytes32Size = FixedSizeUnitDataReadSize
+	EncodedBytes4Size  = FixedSizeUnitDataReadSize
+	EncodedBytes8Size  = FixedSizeUnitDataReadSize
+	EncodedUint64Size  = FixedSizeUnitDataReadSize
+	EncodedUint256Size = FixedSizeUnitDataReadSize
+)
+
+var ErrInputDataTooSmall = errors.New("input data is too small for decoding")
+var ErrBufferTooSmall = errors.New("buffer too small for encoding")
+var ErrDataTooLarge = errors.New("input data is too large for encoding")
+
+// ReadAddress reads an address from the buffer at index
+func ReadAddress(buffer []byte, index int) (gethCommon.Address, error) {
+	if len(buffer) < index+FixedSizeUnitDataReadSize {
+		return gethCommon.Address{}, ErrInputDataTooSmall
+	}
+	paddedData := buffer[index : index+FixedSizeUnitDataReadSize]
+	// addresses are zero-padded on the left side.
+	addr := gethCommon.BytesToAddress(
+		paddedData[FixedSizeUnitDataReadSize-gethCommon.AddressLength:])
+	return addr, nil
+}
+
+// EncodeBool encodes boolean into fixed size unit of encoded data
+func EncodeAddress(address gethCommon.Address, buffer []byte, index int) error {
+	if len(buffer) < index+EncodedAddressSize {
+		return ErrBufferTooSmall
+	}
+	copy(buffer[index:index+EncodedAddressSize],
+		gethCommon.LeftPadBytes(address[:], EncodedAddressSize))
+	return nil
+}
+
+// ReadBool reads a boolean from the buffer at the index
+func ReadBool(buffer []byte, index int) (bool, error) {
+	if len(buffer) < index+EncodedBoolSize {
+		return false, ErrInputDataTooSmall
+	}
+	// bools are zero-padded on the left side
+	// so we only need to read the last byte
+	return uint8(buffer[index+EncodedBoolSize-1]) > 0, nil
+}
+
+// EncodeBool encodes a boolean into fixed size unit of encoded data
+func EncodeBool(bitSet bool, buffer []byte, index int) error {
+	if len(buffer) < index+EncodedBoolSize {
+		return ErrBufferTooSmall
+	}
+	var bitSetVar uint8
+	if bitSet {
+		bitSetVar = 1
+	}
+	copy(buffer[index:index+EncodedBoolSize],
+		gethCommon.LeftPadBytes(
+			[]byte{bitSetVar},
+			EncodedBoolSize,
+		),
+	)
```

Comment:
nit but wouldn't this be more efficient than copying:
```go
    for i := 0; i < EncodedBoolSize; i++ {
        buffer[index+i] = 0
    }

    if bitSet {
        buffer[index+EncodedBoolSize-1] = 1
    }
```

Reply:
yes, good suggestion.